Sources/Purchasing/StoreKit2/Observer Mode/StoreKit2ObserverModePurchaseDetector.swift

```diff
@@ -0,0 +1,124 @@
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  StoreKit2ObserverModePurchaseDetector.swift
+//
+//  Created by Will Taylor on 5/1/24.
+
+import Foundation
+import StoreKit
+
+/// A delegate protocol for handling verified transactions in observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+// swiftlint:disable type_name
+protocol StoreKit2ObserverModePurchaseDetectorDelegate: AnyObject, Sendable {
+
+    /// Handles a verified transaction with its corresponding JWS representation.
+    /// - Parameters:
+    ///   - verifiedTransaction: The verified transaction to be processed.
+    ///   - jwsRepresentation: The JSON Web Signature representation of the transaction.
+    func handleSK2ObserverModeTransaction(
+        verifiedTransaction: StoreKit.Transaction,
+        jwsRepresentation: String
+    ) async throws
+}
+
+/// Protocol describing an actor capable of detecting purchases from StoreKit 2.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+protocol StoreKit2ObserverModePurchaseDetectorType: Actor {
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModePurchaseDetectorDelegate?
+    ) async
+}
+
+/// Actor responsibile for detecting purchases from StoreKit2 that should be processed by observer mode.
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+actor StoreKit2ObserverModePurchaseDetector: StoreKit2ObserverModePurchaseDetectorType {
+
+    private let deviceCache: DeviceCache
+    private let currentUserProvider: CurrentUserProvider
+    private let allTransactionsProvider: AllTransactionsProviderType
+
+    init(
+        deviceCache: DeviceCache,
+        currentUserProvider: CurrentUserProvider,
+        allTransactionsProvider: AllTransactionsProviderType
+    ) {
+        self.deviceCache = deviceCache
+        self.currentUserProvider = currentUserProvider
+        self.allTransactionsProvider = allTransactionsProvider
+    }
+
+    /// Detects unobserved transactions and forwards them to the StoreKit2ObserverModeManagerDelegate
+    /// for processing.
+    func detectUnobservedTransactions(
+        delegate: StoreKit2ObserverModePurchaseDetectorDelegate?
+    ) async {
+        let allTransactions = await allTransactionsProvider.getAllTransactions()
+        guard let mostRecentTransaction = self.getMostRecentVerifiedTransaction(from: allTransactions) else {
+            return
+        }
+
+        let jwsRepresentation = mostRecentTransaction.jwsRepresentation
+        guard let transaction = mostRecentTransaction.verifiedTransaction else { return }
+
+        var cachedSyncedSK2ObserverModeTransactionIDs = Set(
+            self.deviceCache.cachedSyncedSK2ObserverModeTransactionIDs(
+                appUserID: currentUserProvider.currentAppUserID
+            ) ?? []
+        )
+
+        if cachedSyncedSK2ObserverModeTransactionIDs.contains(transaction.id) { return }
+
+        do {
```

Comment:
> This only handles the case where there is a single unknown transaction, and it is the latest one, right?

Yeah, that's right! The idea is that when it's sent up to the backend, the backend will pick up any other unprocessed transactions when it refreshes the receipt. That way, even if a transaction is skipped over by observer mode, it's still detected by the backend.

If we were to send up all unprocessed transactions, that'd probably result in some unnecessary network requests.

If we wanted to loosen this up a bit, we could make it send up the most recent unprocessed transaction, even if it isn't the newest in the queue instead of only transactions that are the most recent transaction in the queue **and** unprocessed.